src/react/components/MappingEditor.tsx

```diff
@@ -0,0 +1,333 @@
+/*
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import React from 'react';
+import { Col, Form, InputGroup } from 'react-bootstrap';
+import { AstarteMapping } from 'astarte-client';
+import type { AstarteInterface } from 'astarte-client';
+import _ from 'lodash';
+
+const astarteDataTypes: AstarteMapping['type'][] = [
+  'string',
+  'boolean',
+  'double',
+  'integer',
+  'longinteger',
+  'binaryblob',
+  'datetime',
+  'doublearray',
+  'integerarray',
+  'booleanarray',
+  'longintegerarray',
+  'stringarray',
+  'binaryblobarray',
+  'datetimearray',
+];
+
+const defaultMapping: AstarteMapping = {
+  endpoint: '',
+  type: 'double',
+};
+
+interface Props {
+  interfaceType: AstarteInterface['type'];
+  interfaceAggregation?: AstarteInterface['aggregation'];
+  mapping?: AstarteMapping;
+  onChange: (updatedMapping: AstarteMapping) => unknown;
+}
+
+export default ({
+  interfaceType,
+  interfaceAggregation = 'individual',
+  mapping = defaultMapping,
+  onChange,
+}: Props): React.ReactElement => {
+  const isPropertiesInterface = interfaceType === 'properties';
+  const isDatastreamIndividualInterface =
+    interfaceType === 'datastream' && interfaceAggregation === 'individual';
+  const showMappingExpiry = mapping.retention === 'volatile' || mapping.retention === 'stored';
+  const showInterfaceDatabaseRetentionTtl = mapping.databaseRetentionPolicy === 'use_ttl';
+
+  let mappingValidationErrors: { [property: string]: string } = {};
+  try {
+    AstarteMapping.validation.validateSync(mapping, { abortEarly: false });
+  } catch (err) {
+    mappingValidationErrors = _.mapValues(_.keyBy(_.uniqBy(err.inner, 'path'), 'path'), 'message');
+  }
+
+  return (
```

Comment:
this is a huge render chunk. It would improbe readability to split it in different macro components.
Like `<MappingTypeSelector value={} onChange={} />`
If we change layout to the component, it will be easier to re-arrange things

Reply:
Personally I don't find it to be that bad, it's actually pretty straightforward in terms of readability.
If we were to use macro components to hide details of each form field, I think readability would suffer: with the current structure you can clearly see that how each field behaves and that the same pattern was used to implement them.
We could keep in mind the approach you suggested if we were to complicate things further, but as of now I have to disagree